File: ns_spam/src/cli_request.cpp
#include "cli_request.h"
#include <iostream>
#include <format>

// CLI colours
static const std::string RED     = "\033[31m";
static const std::string GREEN   = "\033[32m";
static const std::string YELLOW  = "\033[33m";
static const std::string BLUE    = "\033[34m";
static const std::string MAGENTA = "\033[35m";
static const std::string CYAN    = "\033[36m";
static const std::string RESET   = "\033[0m";
static const std::string BOLD    = "\033[1m";
static const std::string ORANGE  = "\033[38;5;208m";

CliRequest::CliRequest(int argc, char *argv[]) {
    for (int i = 1; i < argc; ++i) {
        std::string arg = argv[i];

        if (arg == "--nameserver" && i + 1 < argc) {
            nameserver = argv[++i];
        } else if (arg == "--domain" && i + 1 < argc) {
            domain = argv[++i];
        } else if ((arg == "--requests" || arg == "-n") && i + 1 < argc) {
            requests = std::stoi(argv[++i]);
        } else if ((arg == "--threads" || arg == "-t") && i + 1 < argc) {
            threads = std::stoi(argv[++i]);
        } else if (arg == "--verbose" || arg == "-v") {
            verbose = true;
        } else if (arg == "--random" || arg == "-r") {
            random = true;
        } else if (arg == "--timeout" && i + 1 < argc) {
            timeout = std::stoi(argv[++i]);
        } else if (arg != "--cli") {
            std::cerr << std::format("{}Unknown argument: {}{}\n", RED, arg, RESET);
            exit(1);
        }
    }

    if (nameserver.empty()) {
        std::cerr << std::format(
            "{}ERROR: --nameserver is required{}\n",
            RED, RESET
        );
        exit(1);
    }
}

void CliRequest::print_welcome() {
    std::string ascii_banner = std::format(
        "{}{}"
        " _   _  _____       _____ _____        __  __ \n"
        "| \\ | |/ ____|     / ____|  __ \\ /\\   |  \\/  |\n"
        "|  \\| | (___ _____| (___ | |__) /  \\  | \\  / |\n"
        "| . ` |\\___ \\______\\___ \\|  ___/ /\\ \\ | |\\/| |\n"
        "| |\\  |____) |     ____) | |  / ____ \\| |  | |\n"
        "|_| \\_|_____/     |_____/|_| /_/    \\_\\_|  |_|\n\n\n{}",
        BOLD, CYAN, RESET
    );
    std::cout << ascii_banner;

    std::cout << std::format(
        "{}Thank you for using {}NS-Spam{}.\n",
        YELLOW, MAGENTA, RESET
    );

    std::cout << std::format(
        "{}This application should only be run on nameservers you have permission "
        "from the owner to use.{}\n",
        YELLOW, RESET
    );

    std::cout << std::format(
        "{}You have selected nameserver {}{}{}{}{}{}\n",
        YELLOW, BOLD, BLUE, nameserver, RESET, YELLOW, RESET
    );

    if (random) {
        std::cout << std::format(
            "{}Random mode enabled — generating random subdomains under {}{}{}{}\n",
            YELLOW, BLUE, domain, RESET, YELLOW
        );
    } else {
        std::cout << std::format(
            "{}Domain selected: {}{}{}{}\n",
            YELLOW, BLUE, domain, RESET, YELLOW
        );
    }

    std::cout << std::format(
        "{}Requests: {}{}{} | Threads: {}{}{}\n\n",
        YELLOW, BLUE, requests, YELLOW, BLUE, threads, RESET
    );
}

int CliRequest::run() {
    print_welcome();

    std::cout << std::format("{}Testing nameserver and domain...\n{}", YELLOW, ORANGE);

    if (test_valid()) {
        output(std::format("{}Request Successful{}", GREEN, RESET));
    } else {
        std::cerr << std::format(
            "{}Request failed — invalid nameserver or domain{}\n",
            RED, RESET
        );
        return 1;
    }

    std::cout << std::format(
        "\n\n{}Starting workload...{}\n",
        YELLOW, RESET
    );

    start_requests();

    std::cout << std::format(
        "{}Requests completed\n{}",
        GREEN, RESET
    );

    return 0;
}

void CliRequest::output(std::string str) {
    std::cout << str;
}

File: ns_spam_java/src/main/java/ns/CliRequest.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package ns;

/**
 *
 * @author angus
 */
public class CliRequest extends Request {

    public static final String RED     = "\033[31m";
    public static final String GREEN   = "\033[32m";
    public static final String YELLOW  = "\033[33m";
    public static final String BLUE    = "\033[34m";
    public static final String MAGENTA = "\033[35m";
    public static final String CYAN    = "\033[36m";
    public static final String RESET   = "\033[0m";
    public static final String BOLD    = "\033[1m";
    public static final String ORANGE  = "\033[38;5;208m";

    public CliRequest(String[] args) {
        for (int i = 0; i < args.length; i++) {
            switch (args[i]) {
                case "--cli":
                    break;
                case "--nameserver":
                    if (i + 1 < args.length) {
                        this.nameserver = args[i + 1];
                        i++;
                    } else {
                        throw new IllegalArgumentException("Missing value for --nameserver");
                    }
                    break;
                case "--domain":
                case "-d":
                    if (i + 1 < args.length) {
                        this.domain = args[i + 1];
                        i++;
                    } else {
                        throw new IllegalArgumentException("Missing value for --domain");
                    }
                    break;
                case "--number":
                case "-n":
                    if (i + 1 < args.length) {
                        try {
                            this.number = Integer.parseInt(args[i + 1]);
                        } catch (NumberFormatException e) {
                            System.err.printf("Invalid number input. Must be integer. You entered %s\n", args[i + 1]);
                        }
                        i++;
                    } else {
                        throw new IllegalArgumentException("Missing value for --number");
                    }
                    break;
                case "--threads":
                case "-t":
                    if (i + 1 < args.length) {
                        try {
                            this.threads = Integer.parseInt(args[i + 1]);
                        } catch (NumberFormatException e) {
                            System.err.printf("Invalid threads input. Must be integer. You entered %s\n", args[i + 1]);
                        }
                        i++;
                    } else {
                        throw new IllegalArgumentException("Missing value for --threads");
                    }
                    break;
                case "--timeout":
                    if (i + 1 < args.length) {
                        try {
                            this.timeout = Integer.parseInt(args[i + 1]);
                        } catch (NumberFormatException e) {
                            System.err.printf("Invalid timeout input. Must be integer. You entered %s\n", args[i + 1]);
                            System.exit(1);
                        }
                        i++;
                    } else {
                        throw new IllegalArgumentException("Missing value for --timeout");
                    }
                    break;
                case "--verbose":
                case "-v":
                    this.verbose = true;
                    break;
                case "--random":
                    this.random = true;
                    break;
                case "--endless":
                    this.endless = true;
                    break;
                default:
                    throw new AssertionError();
            }
        }
        if (nameserver == null) {
            System.out.println(RED + "No Nameserver selected. You must select a nameserver." + RESET);
            System.exit(1);
        }

        print_options();
    }

    private void print_options() {
        String asciiBanner = String.format(
            """
            %s%s
                        _   _  _____       _____ _____        __  __ 
                        | \\ | |/ ____|     / ____|  __ \\ /\\   |  \\/  |
                        |  \\| | (___ _____| (___ | |__) /  \\  | \\  / |
                        | . ` |\\___ \\______\\___ \\|  ___/ /\\ \\ | |\\/| |
                        | |\\  |____) |     ____) | |  / ____ \\| |  | |
                        |_| \\_|_____/     |_____/|_| /_/    \\_\\_|  |_|                                   
            %s
            """,
            CYAN, BOLD, RESET
        );

        // Build message
        StringBuilder sb = new StringBuilder();

        sb.append(String.format("%sThank you for using %sNS-Spam%s.\n",
                YELLOW, MAGENTA, RESET));

        sb.append(String.format(
                "%sThis application should only be run on nameservers you have permission from the owner to use.%s\n",
                YELLOW, RESET));

        sb.append(String.format(
                "%sYou have selected nameserver %s%s%s %s to run on.%s\n",
                YELLOW, BOLD, BLUE, nameserver, RESET, YELLOW, RESET));

        if (random) {
            sb.append(String.format(
                    "%sRandom option selected.%s %sThis will create random subdomains off of your domain %s%s%s %s with a length of 3-7 characters.%s\n",
                    YELLOW, RESET, YELLOW, BOLD, BLUE, domain, RESET, YELLOW, RESET
            ));
        } else {
            sb.append(String.format(
                    "%sYou have selected the domain %s%s%s %s to run on.%s\n",
                    YELLOW, BOLD, BLUE, domain, RESET, YELLOW, RESET
            ));
        }

        System.out.println(asciiBanner);
        System.out.println(MAGENTA + "-".repeat(114) + RESET);
        System.out.println(sb);
        System.out.println(MAGENTA + "-".repeat(114) + RESET);
        System.out.println("\n" + YELLOW + "Press enter to begin or CTRL+C to exit" + RESET);

        Thread shutdownHook = new Thread(() -> {
            System.out.println("\n" + GREEN + "Exiting gracefully (Ctrl+C detected)" + RESET);
        });
        Runtime.getRuntime().addShutdownHook(shutdownHook);

        try {
            System.in.read();
        } catch (Exception e) {
            System.out.println("\n" + GREEN + "Exiting gracefully" + RESET);
            System.exit(0);
        } finally {
            try {
                Runtime.getRuntime().removeShutdownHook(shutdownHook);
            } catch (IllegalStateException ignored) {
            }
        }
    }

}
